src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,220 @@
+                        static (arg, cancellationToken) => arg.@this.ComputeHandlersAsync(arg.messageName, arg.isSolution, cancellationToken),
+                        (messageName, arg.@this, arg.isSolution)),
+                    (messageName, @this: this, isSolution));
+            }
+
+            var handlers = await lazyHandlers.GetValueAsync(cancellationToken).ConfigureAwait(false);
+            if (handlers.Length == 0)
+                throw new InvalidOperationException($"No handler found for message {messageName}.");
+
+            if (handlers.Length > 1)
+                throw new InvalidOperationException($"Multiple handlers found for message {messageName}.");
+
+            var handler = handlers[0];
+            if (s_disabledExtensionHandlers.TryGetValue(handler, out _))
+                throw new InvalidOperationException($"Handler was disabled due to previous exception.");
+
+            try
+            {
+                var message = JsonSerializer.Deserialize(jsonMessage, handler.MessageType);
+                var result = await ((IExtensionMessageHandlerWrapper<TArgument>)handler).ExecuteAsync(
+                    message, executeArgument, cancellationToken).ConfigureAwait(false);
+                return JsonSerializer.Serialize(result, handler.ResponseType);
+            }
+            catch (Exception ex) when (DisableHandlerAndPropagate(ex))
+            {
+                throw ExceptionUtilities.Unreachable();
+            }
+
+            bool DisableHandlerAndPropagate(Exception ex)
+            {
+                FatalError.ReportNonFatalError(ex, ErrorSeverity.Critical);
+
+                // Any exception thrown in this method is left to bubble up to the extension. But we unregister this handler
+                // from that assembly to minimize the impact.
+                s_disabledExtensionHandlers.TryAdd(handler, handler);
+                return false;
+            }
+        }
+
+        private async Task<ImmutableArray<IExtensionMessageHandlerWrapper>> ComputeHandlersAsync(
+            string messageName, bool isSolution, CancellationToken cancellationToken)
+        {
+            using var _1 = ArrayBuilder<ExtensionFolder>.GetInstance(out var extensionFolders);
+            lock (_gate)
+            {
+                foreach (var (_, extensionFolder) in _folderPathToExtensionFolder_useOnlyUnderLock)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    extensionFolders.Add(extensionFolder);
+                }
+            }
+
+            using var _ = ArrayBuilder<IExtensionMessageHandlerWrapper>.GetInstance(out var result);
+            foreach (var extensionFolder in extensionFolders)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                await extensionFolder.AddHandlersAsync(messageName, isSolution, result, cancellationToken).ConfigureAwait(false);
```

Comment:
Took me a second to think through this, I think it is fine but writing out to confirm.

We aren't holding `_gate` here, meaning an unregister can come through while this is running.  However we have a second gate inside the `extensionFolder` which means the impls won't run at the same time.  And even if we unregister a folder before we call `AddHandlersAsync` on the same folder, it will just find zero handlers.

Reply:
Correct.  :)